pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+			Ok((weight_data.0 .1, weight_data.1 .1))
+		} else if pool_data.assets == (weight_data.1 .0, weight_data.0 .0) {
+			Ok((weight_data.1 .1, weight_data.0 .1))
+		} else {
+			Err(Error::<T>::InvalidAsset.into())
+		}
+	}
+}
+
+pub trait AssetPairPoolIdFor<AssetId: Sized, PoolId: Sized> {
+	fn from_assets(asset_a: AssetId, asset_b: AssetId) -> PoolId;
+}
+
+pub struct AssetPairPoolId<T: Config>(PhantomData<T>);
+
+impl<T: Config> AssetPairPoolIdFor<AssetId, PoolId<T>> for AssetPairPoolId<T>
+where
+	PoolId<T>: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+{
+	fn from_assets(asset_a: AssetId, asset_b: AssetId) -> PoolId<T> {
+		let mut buf: Vec<u8> = b"lbp".to_vec();
+
+		if asset_a < asset_b {
+			buf.extend_from_slice(&asset_a.to_le_bytes());
+			buf.extend_from_slice(&asset_b.to_le_bytes());
+		} else {
+			buf.extend_from_slice(&asset_b.to_le_bytes());
+			buf.extend_from_slice(&asset_a.to_le_bytes());
+		}
+		PoolId::<T>::unchecked_from(T::Hashing::hash(&buf[..]))
+	}
+}
+
+impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, BalanceOf<T>> for Pallet<T> {
+	fn exists(assets: AssetPair) -> bool {
+		let pair_account = T::AssetPairPoolId::from_assets(assets.asset_in, assets.asset_out);
+		<PoolData<T>>::contains_key(&pair_account)
+	}
+
+	fn get_pair_id(assets: AssetPair) -> T::AccountId {
+		T::AssetPairPoolId::from_assets(assets.asset_in, assets.asset_out)
+	}
+
+	fn get_pool_assets(pool_account_id: &T::AccountId) -> Option<Vec<AssetId>> {
+		let maybe_pool = <PoolData<T>>::try_get(pool_account_id);
+		if let Ok(pool_data) = maybe_pool {
+			Some(vec![pool_data.assets.0, pool_data.assets.1])
+		} else {
+			None
+		}
+	}
+
+	/// Calculate spot price for given assets and amount. This method does not modify the storage.
+	///
+	/// Provided assets must exist in the pool. Panic if an asset does not exist in the pool.
+	///
+	/// Return 0 if calculation overflows or weights calculation overflows.
+	fn get_spot_price_unchecked(asset_a: AssetId, asset_b: AssetId, amount: BalanceOf<T>) -> BalanceOf<T> {
+		let pool_id = Self::get_pair_id(AssetPair {
+			asset_in: asset_a,
+			asset_out: asset_b,
+		});
+
+		let asset_a_reserve = T::MultiCurrency::free_balance(asset_a, &pool_id);
+		let asset_b_reserve = T::MultiCurrency::free_balance(asset_b, &pool_id);
+
+		let pool_data = match <PoolData<T>>::try_get(&pool_id) {
+			Ok(pool) => pool,
+			Err(_) => return BalanceOf::<T>::zero(),
+		};
+
+		// calculate actual weights or reuse the last if calculation is not necessary
+		let (weight_in, weight_out) = match Self::get_actual_weights(&pool_data) {
+			Ok(weights) => {
+				if asset_a == pool_data.assets.0 {
+					(weights.0, weights.1)
+				} else {
+					// swap weights if assets are in different order
+					(weights.1, weights.0)
+				}
+			}
+			Err(_) => return BalanceOf::<T>::zero(),
+		};
+
+		hydra_dx_math::lbp::calculate_spot_price(asset_a_reserve, asset_b_reserve, weight_in, weight_out, amount)
```

Comment:
spot price ignores fee? this would be solved by earlier comment

Reply:
transfer fee does not make sense for the spot price